packages/flutter/lib/src/widgets/sliver_persistent_header.dart

```diff
@@ -185,8 +187,65 @@ class SliverPersistentHeader extends StatelessWidget {
   }
 }
 
+class _FloatingHeader extends StatefulWidget {
```

Comment:
can this merge into the _SliverFloatingPersistentHeader and _SliverFloatingPinnedPersistentHeader, probably as mixin? you may need to change the two widgets to statefulwidget or somehow inject the logic into _SliverPersistentHeaderElement.

It feels a bit convoluted that how it is inserted into the widget tree and can only work with floating persistent header.

Reply:
I agree, I tried that first, but then I could not find the `RenderSliverFloatingPersistentHeader` as an ancestor from that point. I'll look again and see if there is another way.

Reply:
you can do something like context.findRenderObject to get the renderobject (which should be RenderSliverFloatingPersistentHeader if there is no additional renderobject in between _SliverFloatingPersistentHeader and RenderSliverFloatingPersistentHeader)

Reply:
so the structure i am imagine is the _SliverFloatingPersistentHeader listens to the scrollable and updates its renderobject(assuming it is RenderSliverFloatingPersistentHeader) based on it.

In terms of how _SliverFloatingPersistentHeader can listen to scrollable, you can either change it to a stateful widget (probably overkill), or makes _SliverFloatingPersistentHeader inject some callback into _SliverPersistentHeaderElement, and figure out how _SliverPersistentHeaderElement can listen to inherited widget

Reply:
I have not figures out how to do this yet, not sure if I am understanding completely. `_SliverFloatingPersistentHeader` is a `RenderObjectWidget`, and `_SliverPersistentHeaderElement` is a `RenderObjectElement`. 
My original plan here in introducing `_FloatingHeader` was to follow to same pattern we use for `SliverAppBar`, which applies a `_FloatingAppBar` in order to listen to the scroll events. Should this be rewritten? It would probably be better for maintainability if we used the same logic one way or the other.

Reply:
Ah we have similar situation in the appbar as well. If we have it here, do we still need the logic in _FloatingAppBar? it looks the scrolling maybe listen twice for SliverAppBar?